### PR TITLE
Add reduce option to F.black_out

### PR DESCRIPTION
**Pull request by delta2323**
This PR adds `reduce` option to `F.black_out`. I also write its specification in more detail. Related to #2558.

### CODE REVIEW

File: chainer/functions/loss/black_out.py [hunk 1]
```diff
@@ -27,14 +27,33 @@ def black_out(x, t, W, samples):
        p(y) = \\frac{\\exp(W_y^\\top x)}{
        \\sum_{s \\in samples} \\exp(W_s^\\top x)}.
 
+    The output is a variable whose value depends on the value of
+    the option ``reduce``. If it is ``'samplewise'``, it holds the
```
Comment (unnonouno) [inline chainer/functions/loss/black_out.py]: Why not using `'no'`?

File: chainer/functions/loss/black_out.py [hunk 2]
```diff
@@ -27,14 +27,33 @@ def black_out(x, t, W, samples):
...
+    samplewise loss values. If it is ``'mean'``, this function takes
+    a mean of loss values.
+
     Args:
         x (~chainer.Variable): Batch of input vectors.
+            Its shape should be :math:`(N, D)`.
```
Comment (unnonouno) [inline chainer/functions/loss/black_out.py]: How about using \`\`(N, D)\`\` ? Its numpy term.

File: chainer/functions/loss/black_out.py [hunk 3]
```diff
@@ -27,14 +27,33 @@ def black_out(x, t, W, samples):
...
+            Its shape should be :math:`(N,)`. Each elements :math:`v`
+            should satisfy :math:`0 \geq v \geq V` or :math:`-1`
+            where :math:`V` is the number of label types.
         W (~chainer.Variable): Weight matrix.
+            Its shape should be :math:`(V, D)`
         samples (~chainer.Variable): Negative samples.
+            Its shape should be :math:`(N, S)` where :math:`S` is
+            the number of negative samples.
+        recude (str): Reduction option. Its value must be either
+        ``'samplewise'`` or ``'mean'``. Otherwise,
```
Comment (unnonouno) [inline chainer/functions/loss/black_out.py]: Do you need to insert indent here?

File: chainer/functions/loss/black_out.py [hunk 4]
```diff
@@ -58,5 +77,8 @@ def black_out(x, t, W, samples):
         reshape.reshape(logz, (batch_size, 1)), neg_y)
     ny = exponential.log(1 - exponential.exp(bneg_y - blogz))
     py = reshape.reshape(pos_y, (batch_size,))
-    loss = py - logz + _sum.sum(ny, axis=1)
-    return -_sum.sum(loss) / batch_size
+    loss = -(py - logz + _sum.sum(ny, axis=1))
+    if reduce == 'mean':
+        return _sum.sum(loss) / batch_size
```
Comment (unnonouno) [inline chainer/functions/loss/black_out.py]: Use `average` function.

File: tests/chainer_tests/functions_tests/loss_tests/test_black_out.py [hunk 5]
```diff
@@ -11,6 +11,10 @@
 from chainer.testing import condition
 
 
+@testing.parameterize(
+    {'reduce': 'mean'},
+    {'reduce': 'samplewise'}
+)
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/loss_tests/test_black_out.py]: Check invalid reduction type

File: chainer/functions/loss/black_out.py (hunk 1 above)
Comment (delta2323) [inline chainer/functions/loss/black_out.py]: Because I first thought it is more meaningful than `'no'`. But I come to think `'no'` is also OK and further, we can align it with other functions. So, I will change it to `'no'`.

File: chainer/functions/loss/black_out.py (hunk 2 above)
Comment (delta2323) [inline chainer/functions/loss/black_out.py]: I forgot it but do we have a policy for which to use? I think it is better to align it with others.

File: chainer/functions/loss/black_out.py (hunk 4 above)
Comment (delta2323) [inline chainer/functions/loss/black_out.py]: Fixed

File: chainer/functions/loss/black_out.py (hunk 2 above)
Comment (unnonouno) [inline chainer/functions/loss/black_out.py]: hum, ok

**Comment by unnonouno** [comment]
LGTM
